x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java

```diff
@@ -1116,79 +1156,97 @@ private void assertApiKeyNotCreated(Client client, String keyName) throws Execut
+    private Tuple<List<CreateApiKeyResponse>, List<Map<String, Object>>> createApiKeys(int noOfApiKeys, TimeValue expiration) {
         return createApiKeys(TEST_SUPERUSER, noOfApiKeys, expiration, "monitor");
     }
 
-    private List<CreateApiKeyResponse> createApiKeys(String user, int noOfApiKeys, TimeValue expiration, String... clusterPrivileges) {
+    private Tuple<List<CreateApiKeyResponse>, List<Map<String, Object>>> createApiKeys(
+        String user, int noOfApiKeys, TimeValue expiration, String... clusterPrivileges) {
         final Map<String, String> headers = Collections.singletonMap("Authorization",
             basicAuthHeaderValue(user, TEST_PASSWORD_SECURE_STRING));
         return createApiKeys(headers, noOfApiKeys, expiration, clusterPrivileges);
     }
 
-    private List<CreateApiKeyResponse> createApiKeys(String owningUser, String authenticatingUser,
-                                                     int noOfApiKeys, TimeValue expiration, String... clusterPrivileges) {
+    private Tuple<List<CreateApiKeyResponse>, List<Map<String, Object>>> createApiKeys(
+        String owningUser, String authenticatingUser, int noOfApiKeys, TimeValue expiration, String... clusterPrivileges) {
         final Map<String, String> headers = Map.of(
             "Authorization", basicAuthHeaderValue(authenticatingUser, TEST_PASSWORD_SECURE_STRING),
             "es-security-runas-user", owningUser);
         return createApiKeys(headers, noOfApiKeys, expiration, clusterPrivileges);
     }
 
-    private List<CreateApiKeyResponse> createApiKeys(Map<String, String> headers,
-                                                     int noOfApiKeys, TimeValue expiration, String... clusterPrivileges) {
+    private Tuple<List<CreateApiKeyResponse>, List<Map<String, Object>>> createApiKeys(
+        Map<String, String> headers, int noOfApiKeys, TimeValue expiration, String... clusterPrivileges) {
         return createApiKeys(headers, noOfApiKeys, "test-key-", expiration, clusterPrivileges);
     }
 
-    private List<CreateApiKeyResponse> createApiKeys(Map<String, String> headers, int noOfApiKeys, String namePrefix,
-                                                     TimeValue expiration, String... clusterPrivileges) {
+    private Tuple<List<CreateApiKeyResponse>, List<Map<String, Object>>> createApiKeys(
+        Map<String, String> headers, int noOfApiKeys, String namePrefix, TimeValue expiration, String... clusterPrivileges) {
+        List<Map<String, Object>> metadatas = new ArrayList<>(noOfApiKeys);
         List<CreateApiKeyResponse> responses = new ArrayList<>();
         for (int i = 0; i < noOfApiKeys; i++) {
             final RoleDescriptor descriptor = new RoleDescriptor("role", clusterPrivileges, null, null);
             Client client = client().filterWithHeader(headers);
+            final Map<String, Object> metadata = ApiKeyTests.randomMetadata();
```

Comment:
This seems weird. Everything else in this method is provided by the caller, but the metadata is randomly generated and returned.

Reply:
I thought about letting callers pass it in, but then realised the callers were just going to pass `ApiKeyTests.randomMetadata()` anyway because there is no place where the metadata needs to be more specific. So instead of changing every call sites, I just added it here as an easy change.

Reply:
It feels like we accumulating debt - we don't really know what we want our test-object-creation strategy to be, so we have a weird mix.
I don't know what the answer is, but I think this is a sign we're doing something wrong.

Reply:
> we don't really know what we want our test-object-creation strategy to be

I think this is true for the `metadata` field, at least for now. But it has a reasonable explanation. The metadata currently really does not do much other than being stored literally and retrieved literally. Therefore a simple random is sufficient for existing tests. We will need a better strategy when the metadata is searchable. When that happens, we will need to craft the metadata more specifically because we need to search for specific things. I think this can wait till we start working on the metadata search part.

The existing `createApiKeys` methods are also aligned with the above reasoning. We currently have five of them:
1. `createApiKeys(int noOfApiKeys, TimeValue expiration)`
2. `createApiKeys(String user, int noOfApiKeys, TimeValue expiration, String... clusterPrivileges)`
3. `createApiKeys(String owningUser, String authenticatingUser, int noOfApiKeys, TimeValue expiration, String... clusterPrivileges)`
4. `createApiKeys(Map<String, String> headers, int noOfApiKeys, TimeValue expiration, String... clusterPrivileges)` 
5. `createApiKeys(Map<String, String> headers, int noOfApiKeys, String namePrefix, TimeValue expiration, String... clusterPrivileges)`

They are listed roughly in the order of delegation (or increasing complexity), i.e. (1) calls (2), (2) call (3), etc. For the more generic variation, say (1), it basically randomly chooses most of the attributes for an API key, while the more specific variation, (5), gets to specify many of the attributes. It is also where the `metadata` is currently randomised because the existing tests do not care about the exact content of metadata. When we start working on the metadata search, I imagine we can add a 6th variation of `createApiKeys` so it takes the metadata as an argument, i.e.:
`createApiKeys(Map<String, String> headers, int noOfApiKeys, String namePrefix, TimeValue expiration, Map<String, Object> metadata, String... clusterPrivileges)` and current variation (5) will delegate its call to (6).

Reply:
I think we have a smell - the ApiKeyIntegTests are a bit weirdly stuck together, and would benefit from having some structured thinking about what they're trying to achieve and how best to do that. 

I think that's a separate issue though.